File: 0127-word-ladder/0127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordSet = set(wordList)

        if endWord not in wordSet:
            return 0

        # Optimizations to the bottom approach

        # precompute every possible pattern

        L = len(beginWord)

        wildcards = defaultdict(list)
        for w in wordList:
            for i in range(L):
                pattern = w[:i] + '*' + w[i+1:]
                wildcards[pattern].append(w)
        
        front, back = {beginWord}, {endWord}
        visited = {beginWord, endWord}

        steps = 2

        while front and back:
            if len(front) > len(back):
                front, back = back, front

            next_front = set()

            for w in front:
                for i in range(L):
                    pattern = w[:i] + "*" + w[i+1:]

                    for word in wildcards[pattern]:
                        if word in back:
                            return steps
                        elif word not in visited:
                            visited.add(word)
                            next_front.add(word)
                    wildcards[pattern].clear()
            front = next_front
            steps += 1
    

        return 0
        # queue = deque([beginWord])
        # seen = set([beginWord])

        # steps = 1
        # while queue:
        #     for _ in range(len(queue)):
        #         cur_word = queue.popleft()

        #         if cur_word == endWord:
        #             return steps
                
        #         for i in range(len(cur_word)):
        #             for c in 'abcdefghijklmnopqrstuvwxyz':
        #                 new_word = cur_word[:i] + c + cur_word[i+1:]

        #                 if new_word in wordSet and new_word not in seen:
        #                     seen.add(new_word)
        #                     queue.append(new_word)
                    
            
        #     steps += 1
        
        # return 0

File: 0127-word-ladder/0127-word-ladder.go
func ladderLength(beginWord string, endWord string, wordList []string) int {

    n := len(wordList)
    wordMap := make(map[string]struct{}, n)

    for _, word := range wordList {
        wordMap[word] = struct{}{}
    }


    if _, exists := wordMap[endWord]; !exists {
        return 0
    }
    // if wordMap[endWord] != struct{}{} {
    //     return 0
    // }

    visited := make(map[string]struct{}, n+1)


    // queue := make([]string, 0, n+1)
    queue := []string{beginWord}

    visited[beginWord] = struct{}{}

    // queue = append(queue, beginWord)

    steps := 1
    for len(queue) > 0 {
        size := len(queue)

        for i := 0; i < size; i++ {
            curWord := queue[0]

            if curWord == endWord {
                return steps
            }
            queue = queue[1:]
            b := []byte(curWord)
            for pos := 0; pos < len(b); pos++ {
                original := b[pos]

                // for _, chr := range "abcdefghijklmnopqrstuvwxyz" {
                for c := byte('a'); c <= byte('z'); c++ {
                    if c == original {
                        continue
                    }
                    // var b strings.Builder

                    // b.WriteString(curWord[:i])
                    // b.WriteRune(chr)
                    // b.WriteString(curWord[i+1:])

                    // newW := b.String()
                    
                    b[pos] = c
                    next := string(b)
                    _, exists := wordMap[next]
                    _, exists2 := visited[next]


                    // if wordMap[newW] == struct{}{} && visited[newW] != struct{}{} {
                    if exists && !exists2 {
                        visited[next] = struct{}{}
                        queue = append(queue, next)
                    }
                }

                b[pos] = original
            }
        }
        steps++
    }

    return 0
}